cdc/cdc.cc

```diff
@@ -64,12 +65,119 @@ using namespace std::chrono_literals;
 
 static logging::logger cdc_log("cdc");
 
-class cdc::cdc_service::impl {
+namespace cdc {
+static schema_ptr create_log_schema(const schema&, std::optional<utils::UUID> = {});
+static schema_ptr create_stream_description_table_schema(const schema&, std::optional<utils::UUID> = {});
+static future<> populate_desc(db_context ctx, const schema& s);
+}
+
+class cdc::cdc_service::impl : service::migration_listener::empty_listener {
     db_context _ctxt;
 public:
     impl(db_context ctxt)
         : _ctxt(std::move(ctxt))
-    {}
+    {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.register_listener(this);
+        }
+    }
+    ~impl() {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.unregister_listener(this);
+        }
+    }
+
+    void on_before_create_column_family(const schema& schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        if (schema.cdc_options().enabled()) {
+            auto& db = _ctxt._proxy.get_db().local();
+            auto logname = log_name(schema.cf_name());
+            if (!db.has_schema(schema.ks_name(), logname)) {
+                // in seastar thread
+                auto log_schema = create_log_schema(schema);
+                auto stream_desc_schema = create_stream_description_table_schema(schema);
+                auto& keyspace = db.find_keyspace(schema.ks_name());
+
+                auto log_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), log_schema, timestamp);
+                auto stream_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), stream_desc_schema, timestamp);
+
+                mutations.insert(mutations.end(), std::make_move_iterator(log_mut.begin()), std::make_move_iterator(log_mut.end()));
+                mutations.insert(mutations.end(), std::make_move_iterator(stream_mut.begin()), std::make_move_iterator(stream_mut.end()));
+            }
+        }
+    }
+
+    void on_before_update_column_family(const schema& new_schema, const schema& old_schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        bool is_cdc = new_schema.cdc_options().enabled();
+        bool was_cdc = old_schema.cdc_options().enabled();
+
+        if (is_cdc != was_cdc || is_cdc) {
```

Comment:
Unless I'm missing something this condition is the same as if (is_cdc || was_cdc) - but your version is harder to understand.

Reply:
It is, but imho this expresses better what we are looking for. Of course, had I really wanted to be fully expressive, I would have put a comment, so shame on me.

Reply:
To be honest, I didn't understand what you were looking for in this condition. If it was just "is_cdc != was_cdc", that's one thing - looking for adding or dropping cdc. But when you added "|| is_cdc" you completely changed this condition. Maybe a comment would have helped understand why you wanted to have this complicated way to represent the simple condition was_cdc || is_cdc :-)

Reply:
fixed